core/drivers/regulator/regulator.c

```diff
@@ -0,0 +1,385 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <config.h>
+#include <drivers/regulator.h>
+#include <initcall.h>
+#include <keep.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/mutex.h>
+#include <kernel/panic.h>
+#include <kernel/pm.h>
+#include <kernel/tee_time.h>
+#include <kernel/thread.h>
+#include <libfdt.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <util.h>
+
+static SLIST_HEAD(, regulator) regulator_device_list =
+	SLIST_HEAD_INITIALIZER(regulator);
+
+static void lock_regulator(struct regulator *regulator)
+{
+	/*
+	 * Regulator operation may occur at runtime and during specific
+	 * system power transition: power off, PM suspend and resume.
+	 * These operate upon fastcall entries, under PSCI services
+	 * execution, where non-secure world is not operational. In these
+	 * cases we cannot take a mutex and will expect the mutex is
+	 * unlocked.
+	 */
+	if (thread_get_id_may_fail() == THREAD_ID_INVALID) {
+		assert(!regulator->lock.state);
+		return;
+	}
+
+	mutex_lock(&regulator->lock);
+}
+
+static void unlock_regulator(struct regulator *regulator)
+{
+	if (IS_ENABLED(CFG_ARM32_core) &&
+	    !IS_ENABLED(CFG_WITH_ARM_TRUSTED_FW) &&
+	    thread_get_id_may_fail() == THREAD_ID_INVALID) {
+		/* Path for PM sequences when with local Monitor */
+		return;
+	}
+
+	mutex_unlock(&regulator->lock);
+}
+
+static TEE_Result set_state(struct regulator *regulator, bool on_not_off)
+{
+	if (!regulator->ops->set_state)
+		return TEE_SUCCESS;
+
+	return regulator->ops->set_state(regulator, on_not_off);
+}
+
+static TEE_Result regulator_refcnt_enable(struct regulator *regulator)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	FMSG("%s", regulator_name(regulator));
+
+	if (regulator->supply) {
+		res = regulator_enable(regulator->supply);
+		if (res)
+			return res;
+	}
+
+	lock_regulator(regulator);
+
+	if (!regulator->refcount) {
+		res = set_state(regulator, true);
+		if (res) {
+			EMSG("regul %s set state failed with %#"PRIx32,
+			     regulator_name(regulator), res);
+
+			unlock_regulator(regulator);
+
+			if (regulator->supply &&
+			    regulator_disable(regulator->supply))
+				panic();
+
+			return res;
+		}
+	}
+
+	regulator->refcount++;
+	if (!regulator->refcount)
+		panic();
+
+	FMSG("%s refcount: %u", regulator_name(regulator), regulator->refcount);
+
+	unlock_regulator(regulator);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result regulator_enable(struct regulator *regulator)
+{
+	assert(regulator);
```

Comment:
how about a non-assert check here? I'm not a real fan of it when it's not related to device-tree or boot-only checks.

Maybe you consider there's no need to do so, you tell me :)

Few occurrences in this file

Reply:
It's not expected to call the API with a NULL reference.

Reply:
I have ensured `assert(regulator)` is called at entry of each regulator framework API functions.